src_lockfree/skiplist.ml

```diff
@@ -0,0 +1,337 @@
+
+(* *)
+
+(** [get_random_height max_height] gives a random value [n] in the range from
+    [1] to [max_height] with the desired distribution such that [n] is twice as
+    likely as [n + 1]. *)
+let rec get_random_height max_height =
+  let m = (1 lsl max_height) - 1 in
+  let x = Random.bits () land m in
+  if x = 1 then
+    (* We reject [1] to get the desired distribution. *)
+    get_random_height max_height
+  else
+    (* We do a binary search for the highest 1 bit.  Techniques in
+
+         Using de Bruijn Sequences to Index a 1 in a Computer Word
+         by Leiserson, Prokop, and Randall
+
+       could perhaps speed this up a bit, but this is likely not performance
+       critical. *)
+    let n = 0 in
+    let n, x = if 0xFFFF < x then (n + 0x10, x lsr 0x10) else (n, x) in
+    let n, x = if 0x00FF < x then (n + 0x08, x lsr 0x08) else (n, x) in
+    let n, x = if 0x000F < x then (n + 0x04, x lsr 0x04) else (n, x) in
+    let n, x = if 0x0003 < x then (n + 0x02, x lsr 0x02) else (n, x) in
+    let n, _ = if 0x0001 < x then (n + 0x01, x lsr 0x01) else (n, x) in
+    max_height - n
+
+(* *)
+
+let[@inline] is_marked = function
+  | Link (Mark _) -> true
+  | Link (Null | Node _) -> false
+
+(* *)
+
+let rec find_path t key prev prev_at_level preds succs level lowest = function
+  | Link Null ->
+      Array.unsafe_set preds level prev_at_level;
+      Array.unsafe_set succs level Null;
+      lowest < level
+      &&
+      let level = level - 1 in
+      let prev_at_level = Array.unsafe_get prev level in
+      find_path t key prev prev_at_level preds succs level lowest
+        (Atomic.get prev_at_level)
+  | Link (Node r as curr_node) as before -> begin
+      let next_at_level = Array.unsafe_get r.next level in
+      match Atomic.get next_at_level with
+      | Link (Null | Node _) as next ->
+          let c = t.compare key r.key in
+          if 0 < c then
+            find_path t key r.next next_at_level preds succs level lowest next
+          else begin
+            Array.unsafe_set preds level (Array.unsafe_get prev level);
+            Array.unsafe_set succs level curr_node;
+            if lowest < level then
+              let level = level - 1 in
+              let prev_at_level = Array.unsafe_get prev level in
+              find_path t key prev prev_at_level preds succs level lowest
+                (Atomic.get prev_at_level)
+            else begin
+              if level = 0 && r.incr != Size.used_once then begin
+                Size.update_once t.size r.incr;
+                r.incr <- Size.used_once
+              end;
+              0 = c
+            end
+          end
+      | Link (Mark r) ->
+          if level = 0 then Size.update_once t.size r.decr;
+          find_path t key prev prev_at_level preds succs level lowest
+            (let after = Link r.node in
+             if Atomic.compare_and_set prev_at_level before after then after
+             else Atomic.get prev_at_level)
```

Comment:
This line made me think quite a bit so I'm leaving a note in hope it may help  someone else (or a future version of myself). 

I was wondering why the recursive call here is made at the current level and not at a higher level as if `curr_node` has higher levels there are also marked but these marked could have been missed by the current `find_path` call.  Also, I made the wrong assumption that these missed marks may cause the function calling `find_path` (`add` for example) to do some costlier operations to remove the marked node.  So I was thinking maybe we can prevent it by looping a bit higher here (basically at the max height of the marked node, if possible). 

But : 
- the `add` function actually does not make an avoidable costlier operation : either the node has been completely removed making a `CAS` fail (line 263) (this is unavoidable as the removal operation can be completed by almost all operations)  or the removal operation is not finished and `add` can finish without any additional work
-  it may actually make thing worst as encountering a marked node may happen repetitively. So  if each time this happen `find_path` goes back up a few levels it may be delayed indefinitely.

Reply:
Interesting.  Hmm... Yes, I wonder whether there might be ways to do these traversals more intelligently.

🤔

I think a reasonable basic assumption is that encountering a marked link is something that should happen relatively rarely.  That is because marked links are only used on nodes to be removed and so usually you would expect a particular marked link to be encountered roughly just once &mdash; unless there is heavy ongoing parallelism &mdash; and even in that case only a single thread will succeed in unlinking the node at each level.  Also, the expected number of links (and therefore the number of marked links) on a node is (less than) two, because of the way the node heights are distributed.   So, it is probably better to optimize with the assumption that marked links are rare.